### ISSUE TITLE
git-lfs ls-files showing different OID states in output depending upon position within working tree

### ISSUE THREAD

**Issue by alexsmac** [git-lfs/git-lfs#3406]
git --version output: `git version 2.19.2`
git-lfs --version output: `git-lfs/2.6.0 (GitHub; darwin amd64; go 1.11.2)`
Client: macOS 10.13.6
Remote: Bitbucket Server v5.11.1

I encountered a strange issue with one LFS-ed git repository whereby the output from `git-lfs ls-files` for the same files shows **different** file OID states, i.e: LFS pointer vs full-object as indicated by the `*` or `-` prefix in the `ls-files` output depending upon where within the working tree `git-lfs ls-files` is invoked.

The repository structure is something like this:

```
<working_tree>
   <data>
     <rainbows>
        <unicorns>
           0ee357d9-bc00-4c78-8738-7debdf909d26.bin
```

with git-lfs tracking all *.bin files.  Initially, the repository is cloned excluding **all** LFS objects, i.e with: `git-lfs clone -X="*" <REPOSITORY_PATH>`.  This works correctly, in that no binary objects are downloaded.  Then after some interrogation of the non-LFS objects, an LFS fetchexclude pattern is set via: `git config --local lfs.fetchexclude "**/*/unicorns/*.bin"`, after which `.git/config` correctly contains a `fetchexclude` line:

```
[lfs]
    fetchinclude =
    fetchexclude = **/*/unicorns/*.bin
```

and a `git-lfs pull` is performed, after which the weird things happen:

1. **All** binary files, including those that match the exclude regex are pulled down.
2. `git-lfs ls-files` when invoked from **working_tree** shows lines like this:

```
(standard input):930fb40000 * data/rainbows/unicorns/0ee357d9-bc00-4c78-8738-7debdf909d26.bin
```
where the `*` implies that git-lfs thinks the file is an LFS pointer.  It isn't, it's actually the full file.

3. However, `git-lfs ls-files` when invoked from **working_tree/data** shows this output:

```
(standard input):930fb40000 - data/rainbows/unicorns/0ee357d9-bc00-4c78-8738-7debdf909d26.bin
```

where the `-` implies that git-lfs thinks the file is a full-object, which is correct in that it is - but incorrect in that it shouldn't have been pulled down by the `git-lfs pull`.

I haven't been able to recreate this issue with other repositories which make extensive use of the workflow above and regex patterns, or with some trivial test cases.  My conclusion is that something went wrong with this repository specifically, but I'm not sure what.

I appreciate this issue isn't going to be easy to diagnose as I can't produce a minimal working example.  However, is there some useful diagnostic information that I can provide that would help?

**Comment by bk2204** [comment]
Thanks for writing in about this.  I think you've pinned down two bugs:

1. We produce `-` whenever we're in a subdirectory of the working tree.
2. We determine the character to print based on whether the file _currently checked out_ is the same size as specified in the appropriate pointer.

I'm able to reproduce these both in our test repository, so they shouldn't be too difficult to fix.

**Comment by ttaylorr** [comment]
@bk2204 did you get a chance to close the loop on this?

**Comment by bk2204** [comment]
@ttaylorr No, unfortunately, I didn't. I think I tagged it `bug` because I didn't have time at the moment to fix it but wanted to make sure we didn't lose track of it.

**Comment by alexsmac** [comment]
@bk2204 further to this I also discovered an additional issue when testing this.  Assuming the same directory structure above, but using *fetchinclude*, i.e. after the initial `git-lfs clone -X="*" <REPOSITORY_PATH>` is run, `.git/config` is updated to:
```
[lfs]
    fetchinclude = **/*/unicorns/*.bin
    fetchexclude =
```
and `git-lfs pull` is run, `data/rainbows/unicorns/0ee357d9-bc00-4c78-8738-7debdf909d26.bin` is *NOT* pulled down.  Despite the fact that: `ls **/*/unicorns/*.bin` would correctly list the file.

However, if instead: `fetchinclude = **/unicorns/*.bin` is used, it works (i.e. `git-lfs pull` downloads the file).  The result of both is the same for `ls <PATTERN>` but only the second form works for `git-lfs`.

In this case, there's no reason to use `**/*/unicorns/*.bin` vs `**/unicorns/*.bin` (in my particular use-case, it's being generated programatically), but it looks indicative of a parsing bug in the `git-lfs` file pattern handling.  Both are legal, and so I would expect both to function equivalently.

**Comment by bk2204** [comment]
I think there's work by @zkry on the wildmatch project to make these patterns match `.gitattributes` better so that the behavior is consistent between the two.

**Comment by bk2204** [comment]
I've captured this as an issue at git-lfs/wildcard#14.

**Comment by zkry** [comment]
I can look into this problem further and try to fix it.

**Comment by bk2204** [comment]
You're definitely welcome to if you'd like.

**Comment by jamesbellastro** [comment]
Hey, I'm still having this issue on Ubuntu. Has this been updated somewhere that wasn't linked?

**Comment by chrisd8088** [comment]
@jamesbellastro — Sorry to hear you're having trouble.

Could you post the output of `git lfs env` along with a simple reproduction case so we can see exactly what you're encountering?
